design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+  name: example-gateway
+spec:
+  gatewayClassName: example-gatewayclass
+  listeners:
+    - protocol: HTTP
+      port: 80
+  ...
+```
+
+The `port` or `protocol` fields are required and specify the network port/protocol pair that Envoy will listen for
+connection requests on. TLS configuration must be included in the Gateway listener when "HTTPS" or "TLS" is specified
+for `protocol`. `certificateRef` of a TLS configuration is a reference to a Secret in the operator's namespace
+containing a TLS certificate and private key. The secret MUST contain the following keys and data:
+
+- tls.crt: certificate file contents
+- tls.key: key file contents
+  
+For example:
+
+```yaml
+kind: Gateway
+...
+listeners:
+- protocol: HTTPS
+  port: 443
+  tls:
+    mode: Terminate
+    certificateRef:
+      kind: Secret
+      group: core
+      name: default-cert
+---
+kind: Secret
+apiVersion: v1
+metadata:
+  name: default-cert
+  namespace: contour-operator
+data:
+  tls.crt: <CERT>
+  tls.key: <KEY>
+```
+
+Gateway contains a `routes` field that specifies a schema for associating routes with the `listener` using Kubernetes
+selectors. A Route is a Service APIs resource capable of servicing a request and allows a user to expose a resource
+(i.e. Service) by an externally-reachable URL, proxy traffic to the resource, and terminate SSL/TLS connections. The
+following example will bind the Gateway to Routes that include label "app: example".The `port` or `protocol` fields are required when
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+spec:
+  gatewayClassName: example-gatewayclass
+  listeners:
+    - protocol: HTTP
+      port: 80
+      routes:
+        kind: HTTPRoute
+        selector:
+          matchLabels:
+            app: example
+---
+kind: HTTPRoute
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-httproute
+  labels:
+    app: example
+spec:
+  rules:
+  - forwardTo:
+    - serviceName: example-svc
+      port: 8080
+```
+
+The `addresses` field specifies one or more network addresses requested for this Gateway. If unspecified, an IP address
+will be automatically assigned. If the specified address is invalid or unavailable, the operator MUST indicate this in
+Gateway status. The following is a Gateway example that specifies IP address "1.2.3.4". The address will be used for
+each Gateway listener:
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+spec:
+  gatewayClassName: example-gatewayclass
+  addresses:
+    - type: IPAddress
+      value: 1.2.3.4
+```
+
+Although Gateway supports multiple address types, only "IPAddress" will be supported by the Contour Operator.
+
+## Implementation Details
+
+### Controller
+
+Create a controller that reconciles the GatewayClass resource and performs the following:
+
+- Only reconcile a GatewayClass that specifies `controller: projectcontour.io/contour-operator`.
```

Comment:
The operator should look for domains with `projectcontour.io/`, without needing the `contour-operator` bits I think. Is there a good reason to always add that? 

My thought is you could have one GatewayClass be `projectcontour.io/external` and another be `projectcontour.io/internal`.

Reply:
> The operator should look for domains with projectcontour.io/, without needing the contour-operator bits I think. Is there a good reason to always add that?

I believe so. See https://github.com/projectcontour/contour-operator/pull/171#discussion_r552937717.

> My thought is you could have one GatewayClass be projectcontour.io/external and another be projectcontour.io/internal.

The following workflow should support this use-case:
- Create a `GatewayClass` that references a `ContourConfig` containing the external config bits. The default settings should suffice since a Service of type "LoadBalancer" is used for Envoy.
- Create a `GatewayClass` that references a `ContourConfig` containing the internal config bits, i.e network publishing type "Internal" which creates an Envoy Service of type "ClusterIP" instead of type "LoadBalancer".
- Create a `Gateway` that references the external `GatewayClass`.
- Create a `Gateway` that references the internal `GatewayClass`.

You will now have two separate instances of Contour, each managing an Envoy DaemonSet.

Reply:
I guess my initial thought was having the name `contour-operator` in there. You explained that it's because the operator should manage which makes sense. What should we use then for say the quickstart for this field?

Reply:
> What should we use then for say the quickstart for this field?

@stevesloka We can either version this field upon cutting a release, similar to how we manage the image versions, or we can omit the version from the GatewayClass so Contour Operator will always manage `projectcontour.io/contour-operator` GatewayClasses. The latter is similar to the "latest" release approach.

Reply:
Building version management into this fields smells a bit funny to me. Nothing I can concretely say is wrong, but it feels like a problem that should be solved elsewhere, not using string parsing of this field.